File: core/progress_manager.py
import logging
from typing import Dict, Optional, Any

from .utils import ProgressBar

class ProgressManager:
    """管理多个进度条的帮助类"""
    
    def __init__(self, show_progress: bool = True):
        """
        初始化进度条管理器
        
        Args:
            show_progress: 是否显示进度条
        """
        self.show_progress = show_progress
        self.progress_bars: Dict[str, ProgressBar] = {}
    
    def create_progress_bar(self, name: str, total: int, prefix: str, suffix: str = "") -> Optional[ProgressBar]:
        """
        创建并存储一个进度条
        
        Args:
            name: 进度条名称，用于后续引用
            total: 总步数
            prefix: 进度条前缀
            suffix: 进度条后缀
            
        Returns:
            创建的进度条，如果show_progress为False则返回None
        """
        if not self.show_progress:
            return None
            
        progress_bar = ProgressBar(total=total, prefix=prefix, suffix=suffix)
        self.progress_bars[name] = progress_bar
        return progress_bar
    
    def update_progress(self, name: str, current: Optional[int] = None, suffix: Optional[str] = None) -> None:
        """
        更新指定进度条
        
        Args:
            name: 进度条名称
            current: 当前进度
            suffix: 新的后缀文本
        """
        if not self.show_progress or name not in self.progress_bars:
            return
            
        self.progress_bars[name].update(current, suffix)
    
    def finish_progress(self, name: str, suffix: Optional[str] = None) -> None:
        """
        完成指定进度条
        
        Args:
            name: 进度条名称
            suffix: 完成时的后缀文本
        """
        if not self.show_progress or name not in self.progress_bars:
            return
            
        self.progress_bars[name].finish(suffix)
        del self.progress_bars[name]
    
    def close_all_progress_bars(self, suffix: str = "已终止") -> None:
        """
        关闭所有未完成的进度条
        
        Args:
            suffix: 关闭时显示的后缀文本
        """
        if not self.show_progress:
            return
            
        logging.info(f"关闭 {len(self.progress_bars)} 个未完成的进度条...")
        for name, bar in list(self.progress_bars.items()):
            try:
                self.finish_progress(name, suffix)
            except Exception as e:
                logging.warning(f"关闭进度条 '{name}' 时出错: {str(e)}")

File: audio-processor/internal/ui/progress_manager.go
package ui

import (
	"fmt"
	"sync"

	"github.com/ccp-p/asr-media-cli/audio-processor/pkg/utils"
)

// ProgressManager 管理多个进度条
type ProgressManager struct {
	progressBars map[string]*ProgressBar
	mutex        sync.Mutex
	enabled      bool
	terminal     *TerminalManager
}

// 在初始化时启用终端进度条模式
func NewProgressManager() *ProgressManager {
	// 启用终端进度条模式，将日志重定向到文件
	utils.EnableTerminalProgress()

	return &ProgressManager{
		progressBars: make(map[string]*ProgressBar),
		terminal:     GetTerminalManager(),
		enabled:      true,
	}
}

// CreateProgressBar 创建并注册一个新的进度条
func (pm *ProgressManager) CreateProgressBar(id string, total int, prefix string, suffix string) *ProgressBar {
	pm.mutex.Lock()
	defer pm.mutex.Unlock()

	// 如果已经存在同名进度条，先完成它
	if bar, exists := pm.progressBars[id]; exists {
		bar.Complete("已被替换")
	}

	if !pm.enabled {
		return nil
	}
    // 创建新的进度条对象
    bar := NewProgressBar(total, prefix, suffix)
    
	pm.progressBars[id] = bar
	return bar
}

// GetProgressBar 获取已存在的进度条
func (pm *ProgressManager) GetProgressBar(id string) *ProgressBar {
	pm.mutex.Lock()
	defer pm.mutex.Unlock()

	return pm.progressBars[id]
}

// 在 UpdateProgressBar 方法中使用终端管理器
func (pm *ProgressManager) UpdateProgressBar(id string, progress int, message string) {
	pm.mutex.Lock()
	defer pm.mutex.Unlock()

	// 使用终端管理器更新进度
	bar, exists := pm.progressBars[id]
	if !exists {
		bar = pm.CreateProgressBar(id, 100, "", message)
		pm.progressBars[id] = bar
	}

	bar.Update(progress, message)
	
	pm.terminal.UpdateProgress(fmt.Sprintf("%s | %s", bar.String(), message))
}

// CompleteProgressBar 完成进度条
// CompleteProgressBar 完成进度条
func (pm *ProgressManager) CompleteProgressBar(id string, suffix string) {
    if !pm.enabled {
        return
    }

    pm.mutex.Lock()
    bar, exists := pm.progressBars[id]
    pm.mutex.Unlock()

    if exists && bar != nil { // 添加非空检查
        // 安全设置进度为100%
        if bar.Total <= 0 {
            bar.Total = 100
        }
        bar.Current = bar.Total
        
        bar.Complete(suffix)

        // 完成后移除进度条
        pm.RemoveProgressBar(id)
    }
}

// RemoveProgressBar 移除进度条
func (pm *ProgressManager) RemoveProgressBar(id string) {
	pm.mutex.Lock()
	defer pm.mutex.Unlock()

	delete(pm.progressBars, id)
}

// CloseAll 完成所有进度条
func (pm *ProgressManager) CloseAll(suffix string) {
	if !pm.enabled {
		return
	}

	pm.mutex.Lock()
	bars := make([]*ProgressBar, 0, len(pm.progressBars))
	for _, bar := range pm.progressBars {
		bars = append(bars, bar)
	}
	pm.mutex.Unlock()

	for _, bar := range bars {
		bar.Complete(suffix)
	}

	// 清空进度条映射
	pm.mutex.Lock()
	pm.progressBars = make(map[string]*ProgressBar)
	pm.mutex.Unlock()
}

// PrintStatus 打印当前所有进度条的状态
func (pm *ProgressManager) PrintStatus() {
	if !pm.enabled {
		return
	}

	pm.mutex.Lock()
	defer pm.mutex.Unlock()

	fmt.Println("\n当前进度状态:")
	for id, bar := range pm.progressBars {
		percent := float64(bar.Current) / float64(bar.Total) * 100
		fmt.Printf("- %s: %.1f%% (%d/%d) %s\n",
			id, percent, bar.Current, bar.Total, bar.Suffix)
	}
}
